### PR TITLE
Serialization code for binlog events

### PR DESCRIPTION
**Pull request by fulghum**
Provides support for serializing all Dolt data types into MySQL's binary encoding used in binlog events. Vitess provides good support for deserializing binary values from binlog events into Go datatypes, but doesn't provide any support for serializing types into MySQL's binary format. This PR pulls data out of Dolt's storage system and encodes it into MySQL's binary format. It would be interesting to split out the Dolt storage system specific code and the core MySQL serialization logic in the future, but this seems like the right first step. 

Related to https://github.com/dolthub/dolt/issues/7512

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>908a25c</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>908a25c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>a75b728</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>a75b728</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>f7fb3fb</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>f7fb3fb</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/libraries/doltcore/sqle/binlogreplication/binlog_type_serialization.go [hunk 1]
```diff
@@ -0,0 +1,948 @@
...
+		// both sides of the dot are stored separately.
+		// In this example, we'd have:
+		// - 2 bytes to store the first 3 full digits.
+		// - 4 bytes to store the next 9 full digits.
+		// - 3 bytes to store the 6 fractional digits.
+		precision := decimalType.Precision() // total number of fractional and full digits
+		scale := decimalType.Scale()         // number of fractional digits
+		numFullDigits := precision - scale
+		numFullDigitUint32s := numFullDigits / 9
+		numFractionalDigitUint32s := decimalType.Scale() / 9
```
Comment (jycor) [inline go/libraries/doltcore/sqle/binlogreplication/binlog_type_serialization.go:280]: can use `scale` variable made earlier

File: go/libraries/doltcore/sqle/binlogreplication/binlog_type_serialization.go [hunk 2]
```diff
@@ -0,0 +1,948 @@
...
+		numLeftoverFullDigits := numFullDigits - numFullDigitUint32s*9
+		numLeftoverFractionalDigits := decimalType.Scale() - numFractionalDigitUint32s*9
```
Comment (jycor) [inline go/libraries/doltcore/sqle/binlogreplication/binlog_type_serialization.go:282]: can use `scale` variable made earlier

File: go/libraries/doltcore/sqle/binlogreplication/binlog_json_serialization.go [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+	var keyEntriesBuffer []byte
+	var keysBuffer []byte
+	nextKeysOffset := calculateInitialObjectKeysOffset(len(jsonObject), largeEncoding)
+
+	// Sort the keys so that we can process the keys and values in a consistent order
+	sortedKeys := make([]string, 0, len(jsonObject))
+	for key := range jsonObject {
+		sortedKeys = append(sortedKeys, key)
+	}
+	sort.Strings(sortedKeys)
```
Comment (jycor) [inline go/libraries/doltcore/sqle/binlogreplication/binlog_json_serialization.go:154]: Not sure if it makes any difference in replication, but I think MySQL sorts JSON keys by length then alphabetically.

File: go/libraries/doltcore/sqle/binlogreplication/binlog_json_serialization_test.go [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+			name:     "16,383 byte string value",
+			json:     fmt.Sprintf("%q", generateLargeString(16383)),
+			expected: append([]byte{0xc, 0xff, 0x7f}, []byte(generateLargeString(16383))...),
+		},
+		{
+			// String values over 16,383 bytes use a three byte length encoding
+			name:     "16,384 byte string value",
+			json:     fmt.Sprintf("%q", generateLargeString(16384)),
+			expected: append([]byte{0xc, 0x80, 0x80, 0x1}, []byte(generateLargeString(16384))...),
+		},
```
Comment (jycor) [inline go/libraries/doltcore/sqle/binlogreplication/binlog_json_serialization_test.go:81]: Do we need tests here for small/large objects/arrays?

**Review by jycor — APPROVED** [review]
LGTM

File: go/libraries/doltcore/sqle/binlogreplication/binlog_json_serialization_test.go (hunk 4 above)
Comment (fulghum) [reply go/libraries/doltcore/sqle/binlogreplication/binlog_json_serialization_test.go:81]: Yeah, that's a good suggestion. We've got 'em covered in the Vitess round trip tests below, but it's worth grabbing the bytes and adding a quick sanity check test for each here, too.

File: go/libraries/doltcore/sqle/binlogreplication/binlog_json_serialization.go (hunk 3 above)
Comment (fulghum) [reply go/libraries/doltcore/sqle/binlogreplication/binlog_json_serialization.go:154]: oh wow, that's a quirky sort order 😄  I don't think it matters for replication (I've got some integration tests that actually replicate all the data types to MySQL and it seems happy to get them in this order), but that's worth adding a note into the code for.

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>926e664</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>926e664</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Pull request merged**